test/helpers/customError.js

```diff
@@ -1,19 +1,37 @@
 const { config } = require('hardhat');
+const { expect } = require('chai');
 
 const optimizationsEnabled = config.solidity.compilers.some(c => c.settings.optimizer.enabled);
 
 /** Revert handler that supports custom errors. */
-async function expectRevertCustomError(promise, reason) {
+async function expectRevertCustomError(promise, expectedErrorName, args) {
   try {
     await promise;
     expect.fail("Expected promise to throw but it didn't");
   } catch (revert) {
-    if (reason) {
+    if (!Array.isArray(args)) {
+      expect.fail('Expected 3rd array parameter for error arguments');
+    }
+    if (expectedErrorName) {
       if (optimizationsEnabled) {
         // Optimizations currently mess with Hardhat's decoding of custom errors
-        expect(revert.message).to.include.oneOf([reason, 'unrecognized return data or custom error']);
+        expect(revert.message).to.include.oneOf([expectedErrorName, 'unrecognized return data or custom error']);
       } else {
-        expect(revert.message).to.include(reason);
+        const [, error] = revert.message.match(/'(.*)'/);
+        if (!/([A-Z])\w+\(.*\)/g.test(error)) {
+          expect.fail(`Couldn't parse "${error}" as a custom error`);
+        }
+        const [, errorName] = error.match(/(\w+)\(.*\)/);
+        const argMatches = [...error.replace(errorName, '').matchAll(/(0x[0-9A-f]+|-?\d+|\w+)/g)];
```

Comment:
I've never seen `[0-9A-f]` used before. Is that equivalent to `[0-9A-Fa-f]` ? If so, I'd personally favor the second option that I thing is more common.

Reply:
Yes, it's equivalent because it defines a range, and uppercase has a lower identifier than lowercase, so uppercase `A` to lowercase `f` include uppercase `F` and lowercase `a` in the range.

I'm applying your suggestion.